File: src/main/java/io/github/takusan23/clickmanaita/enchant/EnchantRightClickEffectComponent.java
package io.github.takusan23.clickmanaita.enchant;

import io.github.takusan23.clickmanaita.ClickManaitaPlayerEvent;
import net.minecraft.core.component.DataComponentType;
import net.minecraft.world.item.enchantment.ConditionalEffect;
import net.minecraft.world.item.enchantment.effects.EnchantmentEntityEffect;

import java.util.List;
import java.util.function.UnaryOperator;

/**
 * minecraft:hit_block が右クリックじゃ動作しないので、右クリックで発動する hit_block 。
 * クリック板は右クリックなので。
 * <p>
 * エンチャントのエフェクトのトリガーとして clickmanaita:block_right_click が使えます。
 * エフェクトは {@link ClickManaitaEnchantEntityEffect} 参照、クリックで呼び出している処理は {@link ClickManaitaPlayerEvent} 参照。
 */
class EnchantRightClickEffectComponent {

    public static DataComponentType<List<ConditionalEffect<EnchantmentEntityEffect>>> BLOCK_RIGHT_CLICK_EFFECT_COMPONENT = newDataComponent(builder ->
            builder.persistent(ConditionalEffect.codec(EnchantmentEntityEffect.CODEC).listOf())
    );

    /**
     * DataComponentType の Builder を引数のラムダで渡し、build() したものを返す
     */
    private static <T> DataComponentType<T> newDataComponent(UnaryOperator<DataComponentType.Builder<T>> onBuilder) {
        DataComponentType.Builder<T> builder = DataComponentType.builder();
        onBuilder.apply(builder);
        return builder.build();
    }

}

File: src/main/kotlin/io/github/takusan23/clickmanaita/enchant/EnchantRightClickEffectComponent.kt
package io.github.takusan23.clickmanaita.enchant

import net.minecraft.core.component.DataComponentType
import net.minecraft.world.item.enchantment.ConditionalEffect
import net.minecraft.world.item.enchantment.effects.EnchantmentEntityEffect
import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets

/**
 * minecraft:hit_block が右クリックじゃ動作しないので、右クリックで発動する hit_block 。
 * クリック板は右クリックなので。
 *
 * ```json
 * "effects": {
 *   "clickmanaita:block_right_click": [
 *     {
 *       "effect": { ... }
 *     }
 *   ]
 * },
 * ```
 *
 * エンチャントのエフェクトのトリガーとして clickmanaita:block_right_click が使えます。
 * エフェクトは[ClickManaitaEnchantEntityEffect]参照。
 *
 * クリック時に呼び出している処理は[ClickManaitaEnchantClickCallback]です。エフェクトをトリガーしてます。
 */
object EnchantRightClickEffectComponent {

    /** [net.minecraft.enchantment.Enchantment.getEffect]の引数としてこれを使う */
    val BLOCK_RIGHT_CLICK_EFFECT_COMPONENT: DataComponentType<List<ConditionalEffect<EnchantmentEntityEffect>>> = DataComponentType.builder<List<ConditionalEffect<EnchantmentEntityEffect>>>().apply {
        persistent(ConditionalEffect.codec(EnchantmentEntityEffect.CODEC, LootContextParamSets.HIT_BLOCK).listOf())
    }.build()

}